evm/src/main/java/org/hyperledger/besu/evm/operation/SelfDestructEIP6780Operation.java

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package org.hyperledger.besu.evm.operation;
+
+import org.hyperledger.besu.datatypes.Address;
+import org.hyperledger.besu.datatypes.Wei;
+import org.hyperledger.besu.evm.EVM;
+import org.hyperledger.besu.evm.account.Account;
+import org.hyperledger.besu.evm.account.MutableAccount;
+import org.hyperledger.besu.evm.frame.ExceptionalHaltReason;
+import org.hyperledger.besu.evm.frame.MessageFrame;
+import org.hyperledger.besu.evm.gascalculator.GasCalculator;
+import org.hyperledger.besu.evm.internal.Words;
+
+/** The Self destruct operation that observes EIP-6780 semantics. */
+public class SelfDestructEIP6780Operation extends AbstractOperation {
+
+  /**
+   * Instantiates a new Self destruct operation.
+   *
+   * @param gasCalculator the gas calculator
+   */
+  public SelfDestructEIP6780Operation(final GasCalculator gasCalculator) {
+    super(0xFF, "SELFDESTRUCT", 1, 0, gasCalculator);
+  }
+
+  @Override
+  public OperationResult execute(final MessageFrame frame, final EVM evm) {
+    final Address recipientAddress = Words.toAddress(frame.popStackItem());
+
+    // because of weird EIP150/158 reasons we care about a null account, so we can't merge this.
+    final Account recipientNullable = frame.getWorldUpdater().get(recipientAddress);
+    final Wei inheritance = frame.getWorldUpdater().get(frame.getRecipientAddress()).getBalance();
+
+    final boolean accountIsWarm =
+        frame.warmUpAddress(recipientAddress) || gasCalculator().isPrecompile(recipientAddress);
+
+    final long cost =
+        gasCalculator().selfDestructOperationGasCost(recipientNullable, inheritance)
+            + (accountIsWarm ? 0L : gasCalculator().getColdAccountAccessCost());
+
+    if (frame.isStatic()) {
+      return new OperationResult(cost, ExceptionalHaltReason.ILLEGAL_STATE_CHANGE);
+    } else if (frame.getRemainingGas() < cost) {
+      return new OperationResult(cost, ExceptionalHaltReason.INSUFFICIENT_GAS);
+    }
+
+    final Address address = frame.getRecipientAddress();
+    final MutableAccount account = frame.getWorldUpdater().getAccount(address).getMutable();
+
+    if (account.isNewAccount()) {
```

Comment:
Would prefer to pass in a flag to enable this extra check. Had to compare this and the `SelfDestructEIP6780Operation` to see that this extra if was the only difference. That and there is a bit of duplication between the two.

Reply:
done